## Abstract {.page_break_before}

Existing guidelines in statistical modeling for genomics hold that simpler models have advantages over more complex ones.
Potential advantages include cost, interpretability, and improved robustness across datasets or biological contexts.
In cancer transcriptomics, this manifests as a preference for small "gene signatures", or groups of genes whose expression is used to define cancer subtypes or suggest therapeutic interventions.
In order to test the gene signature robustness assumption directly, we quantified generalization of mutation status prediction models across datasets (from cell lines to human tumors, and vice-versa), and across contexts (holding out entire cancer types from pan-cancer data).
We compared two simple procedures for model selection, one that exclusively relies on cross-validation performance and one that combines cross-validation performance with model sparsity, and we did not observe an overall generalization advantage for the procedure incorporating sparsity information.
This result held across multiple generalization problems, and across both linear models (LASSO logistic regression) and non-linear models (fully connected neural network).
In machine learning applications to cancer transcriptomics data where robustness across biological contexts is the primary goal, we recommend choosing the model that performs best on held-out data or in cross-validation.



