# Introduction {.page_break_before}


## Results {.page_break_before}

### Measuring generalization using open cancer genomics data

Figure 1: graphical explanation of experimental design

### Generalization across datasets

Figure 2: summary of TCGA -> CCLE and CCLE -> TCGA results, by gene

Figure 3: explanation of "best" vs. "smallest good", distribution plot, examples

### Generalization across cancer types

Figure 4: cancer type generalization distribution plot, examples, cancer type representation analysis (which cancer types are hardest to generalize to?)

### Extension to nonlinear models

Figure 5: neural network results for varying dropout/other regularization

supplement: MSI prediction results, LASSO penalty vs. feature count plots, individual per-gene performance plots, feature count binning results


## Discussion {.page_break_before}


## Methods {.page_break_before}

### Mutation data download and preprocessing

To generate binary mutated/non-mutated gene labels for our machine learning model, we used mutation calls for TCGA samples from MC3 [@doi:10.1016/j.cels.2018.03.002] and copy number threshold calls from GISTIC2.0 [@doi:10.1186/gb-2011-12-4-r41].
MC3 mutation calls were downloaded from the Genomic Data Commons (GDC) of the National Cancer Institute, at https://gdc.cancer.gov/about-data/publications/pancanatlas.
Thresholded copy number calls are from an older version of the GDC data and are available here: https://figshare.com/articles/dataset/TCGA_PanCanAtlas_Copy_Number_Data/6144122.
We removed hypermutated samples, defined as two or more standard deviations above the mean non-silent somatic mutation count, from our dataset to reduce the number of false positives (i.e., non-driver mutations).
After this filtering, 9074 TCGA samples with mutation and copy number data remained.
Any sample with either a non-silent somatic variant or a copy number variation (copy number gain in the target gene for oncogenes and copy number loss in the target gene for tumor suppressor genes) was included in the positive set; all remaining samples were considered negative for mutation in the target gene.

We followed a similar procedure to generate binary labels for cell lines from CCLE, using the data available on the DepMap download portal at https://depmap.org/portal/download/all/.
Mutation information was retrieved from the OmicsSomaticMutations.csv data file, and copy number inforamtion was retrieved from the OmicsCNGene.csv data file.
We thresholded the CNV log-ratios provided by CCLE into binary gain/loss calls using a lower threshold of log~2~(3/2) (i.e. cell lines with a log-ratio below this threshold were considered to have a full copy loss in the corresponding gene), and an upper threshold of log~2~(5/2) (i.e. cell lines with a log-ratio above this threshold were considered to have a full copy gain in the corresponding gene).
After applying the same hypermutation criteria that we used for TCGA, no cell lines in CCLE were identified as hypermutated.
After preprocesing, 1402 cell lines with mutation and copy number data remained.
We then combined non-silent point mutations and copy number gain/loss information into binary labels using the same criteria as for TCGA.

### Gene expression data download and preprocessing

### Cancer gene set construction

Vogelstein gene set

### Classifier setup and cross-validation design

LASSO logistic regression details
LASSO range selection

### "Best model" vs. "smallest good model" analysis details

Description of "smallest good" heuristic
Statistical testing?

### Open science/reproducibility stuff
